File: intervals_processor/hill_detector.py
import sys
import os
sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from intervals_processor.base_detector import BaseDetector

class HillDetector(BaseDetector):
    def __init__(self, min_elevation_gain=5.0, min_grade=2.0, min_warmup_dist_m=1000, **kwargs):
        super().__init__(**kwargs)
        self.min_elevation_gain = min_elevation_gain
        self.min_grade = min_grade
        self.min_warmup_dist_m = min_warmup_dist_m
        
    def _detect_blocks(self, processed_dict):
        times = sorted(processed_dict.keys())
        blocks = []
        current_block = []
        
        gap_counter = 0
        max_gap = 5
        
        for t in times:
            data = processed_dict[t]
            grade = data.get("grade_percent", 0)
            vertical_speed = data.get("vertical_speed_m_s", 0)
            
            is_uphill = grade > 1.0 or vertical_speed > 0.05
            
            if is_uphill:
                current_block.append((t, data))
                gap_counter = 0
            else:
                if current_block and gap_counter < max_gap:
                    current_block.append((t, data))
                    gap_counter += 1
                elif current_block:
                    real_block = current_block[:-gap_counter] if gap_counter > 0 else current_block
                    if self._is_valid_hill(real_block):
                        blocks.append(real_block)
                    current_block = []
                    gap_counter = 0
                    
        if current_block:
            real_block = current_block[:-gap_counter] if gap_counter > 0 else current_block
            if self._is_valid_hill(real_block):
                blocks.append(real_block)
        
        # filter warmup distance
        return [b for b in blocks if b[0][1]["distance_total_m"] >= self.min_warmup_dist_m]
    
    def _is_valid_hill(self, block):
        if not block:
            return False
        
        elevation_gain = block[-1][1]["elevation_m"] - block[0][1]["elevation_m"]
        
        distance = block[-1][1]["distance_total_m"] - block[0][1]["distance_total_m"]
        
        avg_grade = (elevation_gain / distance * 100) if distance > 0 else 0
        
        return elevation_gain >= self.min_elevation_gain and avg_grade >= self.min_grade
    
    def _summarize_effort(self, block, label):
        summary = self._summarize_common(block, label)
        elev_gain = block[-1][1]["elevation_m"] - block[0][1]["elevation_m"]
        avg_grade_percent = elev_gain / summary["distance_m"] * 100 if summary["distance_m"] > 0 else 0
        vam = elev_gain / summary["moving_duration_sec"] * 3600 if summary["moving_duration_sec"] > 0 else 0
        
        summary.update({
            "type": label.replace("WORKOUT", "HILL_REPEATS"),
            "elev_gain_m": round(elev_gain, 1),
            "avg_grade_percent": round(avg_grade_percent, 1),
            "vam": round(vam)
        })
        return summary

File: server/src/strava/sync/detectors/hill-detector.ts
import { ProcessedSecond } from '../strava-sync.service';
import { BaseDetector, DetectedLap, ProcessedDict } from './base-detector';

export class HillDetector extends BaseDetector {
  private readonly minElevationGain: number;
  private readonly minGrade: number;
  private readonly minWarmupDistM: number;

  constructor(
    minElevationGain    = 5.0,
    minGrade            = 2.0,
    minWarmupDistM      = 1000,
    minSpeedMoving      = 0.3,
    cooldownSpeedThreshold = 2.2,
  ) {
    super(minSpeedMoving, cooldownSpeedThreshold);
    this.minElevationGain = minElevationGain;
    this.minGrade         = minGrade;
    this.minWarmupDistM   = minWarmupDistM;
  }

  protected detectBlocks(dict: ProcessedDict): ProcessedSecond[][] {
    const times  = Array.from(dict.keys()).sort((a, b) => a - b);
    const blocks: ProcessedSecond[][] = [];
    let currentBlock: ProcessedSecond[] = [];
    let gapCounter = 0;
    const maxGap = 5;

    for (const t of times) {
      const data          = dict.get(t)!;
      const grade         = data.gradePercent ?? 0;
      const verticalSpeed = data.verticalSpeedMs ?? 0;

      const isUphill = grade > 1.0 || verticalSpeed > 0.05;

      if (isUphill) {
        currentBlock.push(data);
        gapCounter = 0;
      } else {
        if (currentBlock.length && gapCounter < maxGap) {
          currentBlock.push(data);
          gapCounter++;
        } else if (currentBlock.length) {
          const realBlock =
            gapCounter > 0 && gapCounter < currentBlock.length
              ? currentBlock.slice(0, -gapCounter)
              : currentBlock;

          if (this.isValidHill(realBlock)) {
            blocks.push(realBlock);
          }

          currentBlock = [];
          gapCounter   = 0;
        }
      }
    }

    if (currentBlock.length) {
      const realBlock =
        gapCounter > 0 && gapCounter < currentBlock.length
          ? currentBlock.slice(0, -gapCounter)
          : currentBlock;

      if (this.isValidHill(realBlock)) {
        blocks.push(realBlock);
      }
    }

    return blocks.filter(
      (b) => b[0].distanceTotalM >= this.minWarmupDistM,
    );
  }

  private isValidHill(block: ProcessedSecond[]): boolean {
    if (!block.length) return false;

    const elevGain =
      block[block.length - 1].elevationM - block[0].elevationM;
    const distance =
      block[block.length - 1].distanceTotalM - block[0].distanceTotalM;
    const avgGrade = distance > 0 ? (elevGain / distance) * 100 : 0;

    return elevGain >= this.minElevationGain && avgGrade >= this.minGrade;
  }

  protected summarizeEffort(
    block: ProcessedSecond[],
    label: string,
  ): DetectedLap {
    const summary     = this.summarizeCommon(block, label);
    const elevGain    = block[block.length - 1].elevationM - block[0].elevationM;
    const distance    = summary.distanceM;
    const movingSec   = summary.movingDurationSec;

    const avgGradePercent =
      distance > 0 ? (elevGain / distance) * 100 : 0;
    const vam =
      movingSec > 0 ? (elevGain / movingSec) * 3600 : 0;

    summary.elevGainM       = Math.round(elevGain * 10) / 10;
    summary.avgGradePercent = Math.round(avgGradePercent * 10) / 10;
    summary.vam             = Math.round(vam);

    return summary;
  }
}